### PR TITLE
experimental recursive CTE

### PR DESCRIPTION
**Pull request by melanieplageman**
This PR brings in experimental support of `WITH RECURSIVE` for CTE, as a follow-up to 41c3b698616bd984b64629c5bae9c0549f07b8ff , where we converged our implementation with upstream.

The main constraint of bringing in `WITH RECURSIVE` is the single-process nature of `Recursive Union` and `Work Table Scan`. We would have to either 1) Introduce a major redesign of cross-slice communication between `Work Table Scan` and `Recursive Union` or 2) Bend over backwards to generate plans that don't put motions along the path from `Work Table Scan`s to their corresponding `Recursive Union`s.

Consequently, `UNION ALL` is the only set operation supported between the recursive and non-recursive term.

Cherry-picking those commits also hopefully helps impending work on 6.0 by reducing the divergence.

While recursive CTE is under development, the functionality is hidden behind a GUC, `gp_recursive_cte_prototype`

### CODE REVIEW

**Comment by melanieplageman** [comment]
This supersedes #2986

**Comment by ivannovick** [comment]
Great to see the progress :)

File: src/backend/parser/parse_cte.c [hunk 1]
```diff
@@ -953,3 +1006,56 @@ checkWellFormedSelectStmt(SelectStmt *stmt, CteState *cstate)
...
+			SelectStmt *subquery = (SelectStmt *) rs->subquery;
+			checkWellFormedSelectStmt(subquery, cstate);
+		}
+	}
+	cstate->context = cxt;
+}
+
+/*
+ * Check if the recursive term of a recursive cte contains a window function.
+ * This is currently not supported and is checked for in the parsting stage
```
Comment (danielgustafsson) [inline src/backend/parser/parse_cte.c]: s/parsting/parsing/

**Comment by cjcjameson** [comment]
Just reviewed the rebase / conflict resolution -- good to get in before finalizing a 5.0.0 so @vraghavan78 please keep track of this to make sure it gets in whatever you tag.

**Comment by d** [comment]
Concourse is acting up a bit. I'll hold off merging and use this down-time as an opportunity to rebase this branch one last time before merging.

**Comment by d** [comment]
trigger

**Comment by d** [comment]
trigger yo

**Pull request merged**